### ISSUE TITLE
Custom plug-in - Elementor\\Controls_Stack::sanitize_settings() error

### ISSUE THREAD

**Issue by euene** [elementor/elementor#16522]
## Prerequisites

- [x] I have searched for similar issues in both open and closed tickets and cannot find a duplicate.
- [x] The issue still exists against the latest stable version of Elementor.

## Description

Trying to develop a custom plug-in that extends the default Nav Menu widget by adding a few extra controls.
However, when trying to get the value of the currently selected Menu control using get_raw_data() to use the value in another control, there's a fatal error with Controls_Stack::sanitize_settings() which breaks the Elementor editor.

## Steps to reproduce

Plug-in example code:

```php
add_action('elementor/element/nav-menu/section_layout/after_section_end', function($element, $args) {
    // Get the selected menu

    // THIS LINE TRIGGERS THE ERROR
    $selected_menu = $element->get_raw_data()['settings']['menu'];

    // Create new section "Test"
    $element->start_controls_section(
        'content_section',
        [
            'label' => 'Test',
            'tab' => \Elementor\Controls_Manager::TAB_CONTENT,
        ]
    );

    $menu = wp_get_nav_menu_items('test-menu');

    foreach ($menu as $item ) {
        $menu_array[$item->ID] = $item->title;
    }

    $element->add_control(
        'menu_items',
        [
            'label' => 'Menu Item',
            'type' => \Elementor\Controls_Manager::SELECT,
            'options' => $menu_array ?? ['no_menu_defined' => 'No menus are defined.'],
        ]
    );

    // End new section "Test"
    $element->end_controls_section();
}, 10, 2);
```

Error on PHP 7.4:

```PHP
PHP Fatal error:  Uncaught TypeError: Argument 1 passed to Elementor\\Controls_Stack::sanitize_settings() must be of the type array, null given, called in wp-content/plugins/elementor/includes/base/controls-stack.php on line 1051 and defined in wp-content/plugins/elementor/includes/base/controls-stack.php:2290\nStack trace:\n#0 wp-content/plugins/elementor/includes/base/controls-stack.php(1051): Elementor\\Controls_Stack->sanitize_settings(NULL)\n#1 wp-content/plugins/elementor/includes/base/element-base.php(532): Elementor\\Controls_Stack->get_data()\n#2 wp-content/plugins/elementor/includes/base/widget-base.php(696): Elementor\\Element_Base->get_raw_data(false)\n#3 wp-content/plugins/test/advanced-menu.php(14): Elementor\\Widget_Base->get_raw_data()\n#4 wp-includes/class-wp-hoo in wp-content/plugins/elementor/includes/base/controls-stack.php on line 2290, referer: http://localhost/wp-admin/post.php?post=211&action=elementor
```

Error on PHP 8.0:

```PHP
PHP Fatal error:  Uncaught TypeError: Elementor\\Controls_Stack::sanitize_settings(): Argument #1 ($settings) must be of type array, null given, called in wp-content/plugins/elementor/includes/base/controls-stack.php on line 1051 and defined in wp-content/plugins/elementor/includes/base/controls-stack.php:2290\nStack trace:\n#0 wp-content/plugins/elementor/includes/base/controls-stack.php(1051): Elementor\\Controls_Stack->sanitize_settings(NULL)\n#1 wp-content/plugins/elementor/includes/base/element-base.php(532): Elementor\\Controls_Stack->get_data()\n#2 wp-content/plugins/elementor/includes/base/widget-base.php(696): Elementor\\Element_Base->get_raw_data(false)\n#3 wp-content/plugins/test/advanced-menu.php(14): Elementor\\Widget_Base->get_raw_data()\n#4 wp-includes/class-wp-hook.php(303): {closure}(Object(ElementorPro\\Modules\\NavMenu\\Widgets\\Nav_Menu), Array)\n#5 wp-includes/class-wp-hook.php(327): WP_Hook->apply_filters('', Array)\n#6 wp-includes/plugin.php(470): WP_Hook->do_action(Array)\n#7 wp-content/plugins/elementor/includes/base/controls-stack.php(1526): do_action('elementor/eleme...', Object(ElementorPro\\Modules\\NavMenu\\Widgets\\Nav_Menu), Array)\n#8 wp-content/plugins/elementor-pro/modules/nav-menu/widgets/nav-menu.php(421): Elementor\\Controls_Stack->end_controls_section()\n#9 wp-content/plugins/elementor/includes/base/controls-stack.php(2205): ElementorPro\\Modules\\NavMenu\\Widgets\\Nav_Menu->register_controls()\n#10 wp-content/plugins/elementor/includes/base/controls-stack.php(480): Elementor\\Controls_Stack->init_controls()\n#11 wp-content/plugins/elementor/includes/base/widget-base.php(172): Elementor\\Controls_Stack->get_stack()\n#12 wp-content/plugins/elementor/includes/managers/widgets.php(274): Elementor\\Widget_Base->get_stack(false)\n#13 [internal function]: Elementor\\Widgets_Manager->ajax_get_widget_types_controls_config(Array, Object(Elementor\\Core\\Common\\Modules\\Ajax\\Module))\n#14 wp-content/plugins/elementor/core/common/modules/ajax/module.php(171): call_user_func(Array, Array, Object(Elementor\\Core\\Common\\Modules\\Ajax\\Module))\n#15 wp-includes/class-wp-hook.php(303): Elementor\\Core\\Common\\Modules\\Ajax\\Module->handle_ajax_request('')\n#16 wp-includes/class-wp-hook.php(327): WP_Hook->apply_filters('', Array)\n#17 wp-includes/plugin.php(470): WP_Hook->do_action(Array)\n#18 wp-admin/admin-ajax.php(187): do_action('wp_ajax_element...')\n#19 {main}\n  thrown in wp-content/plugins/elementor/includes/base/controls-stack.php on line 2290, referer: http://localhost/wp-admin/post.php?post=211&action=elementor
```

## Isolating the problem

- [x] This bug happens with only Elementor plugin active (and Elementor Pro).
- [x] This bug happens with a default WordPress theme active.
- [x] I can reproduce this bug consistently using the steps above.

## Environment

WordPress 5.8.1

Only plug-ins active:
Elementor 3.4.4
Elementor Pro 3.4.1

Active theme:
Hello Elementor 2.4.1

Server:
PHP 8.0.11 and PHP 7.4.24
Apache/2.4.50
MariaDB 10.6.4

**Comment by euene** [comment]
Hi @udidol,

Another issue that I'm having is that the action only runs once, for the first widget instance on the page and will not fire for any subsequent widgets of the same type - therefore I can't the control value of the 2<sup>nd</sup>, 3<sup>rd</sup> etc. widget instance.

```php
add_action('elementor/element/nav-menu/section_layout/after_section_end', function ($element, $args) {
    if (empty($element->get_id())) {
        return;
    }

    // Get the selected menu
    $selected_menu = $element->get_settings('menu');
}, 10, 2);
```

So the value of `$selected_menu` in the 2<sup>nd</sup> or 3<sup>rd</sup> widgets will wrongly have the value of the 1<sup>st</sup> widget.

Is there a workaround for this?

**Comment by udidol** [comment]
Hi @euene,
What are you trying to achieve?

**Comment by euene** [comment]
> Hi @euene, What are you trying to achieve?

@udidol

Same thing that I mentioned earlier in this issue - create a new SELECT control with options based on the value of another control.

```php
add_action('elementor/element/nav-menu/section_layout/after_section_end', function ($element, $args) {
    if (empty($element->get_id())) {
        return;
    }

    // Get the selected menu
    $selected_menu = $element->get_settings('menu');

    // Get menu items of selected menu
    $menu_items = wp_get_nav_menu_items($selected_menu ?? 'fallback-menu-name');

    // Create new array of top level menu items to be used as <select> options in custom control
    if ($menu_items) {
        foreach ($menu_items as $item) {
            if ($item->menu_item_parent == 0) {
                $menu_array[$item->ID] = $item->title;
            }
        }
    }

    // Add the custom control
    $element->start_controls_section(
        'test_section',
        [
            'label' => 'Test',
            'tab' => \Elementor\Controls_Manager::TAB_CONTENT,
        ]
    );

    $element->add_control(
        'menu_items',
        [
            'label' => 'Menu Item',
            'type' => \Elementor\Controls_Manager::SELECT,
            'options' => $menu_array ?? ['no_menu_defined' => 'No menus are defined.'],
        ]
    );

    $element->end_controls_section();
}, 10, 2);
```

Is this possible to do this in PHP with multiple nav menu widgets on the same page that have a different menu selected considering that it appears that the action elementor/element/nav-menu fires only once for the first widget instance?

If not currently possible in PHP, am I correct in assuming that the best workaround would be to check the value of the selected menu via JS,  then populate the options of the second custom SELECT control via AJAX/JS and save the selected value from the custom SELECT control in a third hidden input field custom control?

**Comment by udidol** [comment]
Unfortunately what you are describing is not currently possible via PHP.
If you MUST do this, the JS "hack" you described is theoretically possible.
